File: 188-best-time-to-buy-and-sell-stock-iv/best-time-to-buy-and-sell-stock-iv.c
int maxProfit(int k, int* prices, int pricesSize) {
    // If we can make enough transactions, this becomes
    // the unlimited-transactions problem.
    if (k >= pricesSize / 2) {
        int profit = 0;

        for (int i = 1; i < pricesSize; i++) {
            if (prices[i] > prices[i - 1]) {
                profit += prices[i] - prices[i - 1];
            }
        }

        return profit;
    }

    int buy[k + 1];
    int sell[k + 1];

    for (int t = 0; t <= k; t++) {
        buy[t] = -prices[0];
        sell[t] = 0;
    }

    for (int i = 1; i < pricesSize; i++) {
        int price = prices[i];

        for (int t = 1; t <= k; t++) {
            buy[t] = (buy[t] > sell[t - 1] - price)
                   ? buy[t]
                   : sell[t - 1] - price;

            sell[t] = (sell[t] > buy[t] + price)
                    ? sell[t]
                    : buy[t] + price;
        }
    }

    return sell[k];
}

File: 188-best-time-to-buy-and-sell-stock-iv/best-time-to-buy-and-sell-stock-iv.rs
impl Solution {
    pub fn max_profit(k: i32, prices: Vec<i32>) -> i32 {
        let n = prices.len();
        let k = k as usize;

        // Unlimited transactions case
        if k >= n / 2 {
            let mut profit = 0;

            for i in 1..n {
                if prices[i] > prices[i - 1] {
                    profit += prices[i] - prices[i - 1];
                }
            }

            return profit;
        }

        let mut buy = vec![0; k + 1];
        let mut sell = vec![0; k + 1];

        for t in 1..=k {
            buy[t] = -prices[0];
        }

        for i in 1..n {
            let price = prices[i];

            for t in 1..=k {
                buy[t] = buy[t].max(sell[t - 1] - price);
                sell[t] = sell[t].max(buy[t] + price);
            }
        }

        sell[k]
    }
}